staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go

```diff
@@ -92,9 +93,12 @@ func newETCD3HealthCheck(c storagebackend.Config) (func() error, error) {
 		return true, nil
 	}, wait.NeverStop)
 
-	return func() error {
+	healthzErrorMessage := &atomic.Value{}
+	healthzErrorMessage.Store("etcd client connection not yet established")
+	go wait.Until(func() {
 		if errMsg := clientErrMsg.Load().(string); len(errMsg) > 0 {
-			return fmt.Errorf(errMsg)
+			healthzErrorMessage.Store(errMsg)
+			return
 		}
 		client := clientValue.Load().(*clientv3.Client)
 		healthcheckTimeout := storagebackend.DefaultHealthcheckTimeout
```

Comment:
I don't like that this value now determines how long a cluster with an unhealthy etcd will continue to report that it is healthy. I also don't usually trust things to actually successfully time out-- if it hangs because something not included in the timeout takes forever, then nothing fixes whatever state was last stored.

Reply:
If you don't trust the timeout on etcd client, I can add a hard timeout in the code here directly.

While ensuring we always clear a value eventually is important, I think we can either 
1. trust the etcd code honors the context timeout
2. add a timeout case directly here like we did with aggregated apiserver status checks

But I think that having a constant time healthz check is important.

Reply:
Yeah I think some explicit timeout detecting code here would be good. And some way of noticing if the calls are hanging forever and a cap on how many in parallel we're willing to do (or at least a note that we don't care because the system is unhealthy in that case anyway).